### PR TITLE
Biter improvements

### PR DESCRIPTION
**Pull request by Pacasasgar**
This PR contains the following features/changes:

-Biter explodes while dying. This explosion is 'faster' than the normal explosion and players should move away faster.

-Biter height modified - offset has been reduced.

-Bugfix: Biter crashed when players were dead.

-Bugfix: Biter explosion dealt damage only once. The rest of the explosions weren't dealing damage at all.

-Bugfix: Biter explosion cooldown wasn't working correctly and it was difficult for the Biter to explode. Now it has 10 seconds cooldown (need balancing) and will explode always at melee range. Charge cooldown also modified. Notice that 'delta_time' is being used to track cooldowns instead of 'time' now.

### CODE REVIEW

File: Scripts/src/Script/Biter.cpp [hunk 1]
```diff
@@ -1,396 +1,396 @@
...
+		charge_current_cooldown -= App->time->delta_time;
+	}
+}
+
+void Biter::Explosion()
+{
+	explosion_current_time += App->time->delta_time;
+	owner->transform.SetGlobalMatrixTranslation(current_position + offset);
+	collider->UpdateDimensions();
+}
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:374]: this is called Explosion but this just update the position so, I don't understand this method

File: Scripts/src/Script/Biter.cpp [hunk 2]
```diff
@@ -1,396 +1,396 @@
...
+	//check if any player is alive	
+	if (!GetClosestTarget()) 	
+	{	
+		owner->transform.SetGlobalMatrixTranslation(current_position + offset);	
+		collider->UpdateDimensions();	
+		return;	
+	}	
+	float3 target = GetClosestTarget()->owner->transform.GetGlobalTranslation();
+
+	if (next_position.Equals(float3(0, 0, 0)) || App->time->time > time_count_delay)
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:230]: instead float3(0,0,0) use float3::zero;

File: Scripts/src/Script/Biter.cpp [hunk 3]
```diff
@@ -1,396 +1,396 @@
...
+	float3 desired_direction = current_velocity;
+	Quat desired_rotation = Quat::LookAt(float3::unitZ, desired_direction.Normalized(), float3::unitY, float3::unitY);
+	Quat new_rotation = Quat::Slerp(owner->transform.GetRotation(), desired_rotation, App->time->delta_time * 0.01F);
+	owner->transform.SetRotation(new_rotation);
+
+	owner->transform.SetGlobalMatrixTranslation(current_position + offset);
+	collider->UpdateDimensions();
+}
+
+void Biter::ComputeHorizontalVelocity(float3 direction, float speed)
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:324]: make const float3 direction

File: Scripts/src/Script/Biter.cpp [hunk 4]
```diff
@@ -1,396 +1,396 @@
...
+	explosion_current_time += App->time->delta_time;
+	owner->transform.SetGlobalMatrixTranslation(current_position + offset);
+	collider->UpdateDimensions();
+}
+
+void Biter::BiteAttack()
+{
+	owner->transform.SetGlobalMatrixTranslation(current_position + offset);
+	collider->UpdateDimensions();
+}
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:380]: this method is called BiteAttack but just move the Bitter, so.... check if it's used if not remove it or change the name and if used and change the name make it const

File: Scripts/src/Script/Biter.cpp [hunk 5]
```diff
@@ -1,396 +1,396 @@
...
+	{
+		explosion_current_cooldown -= App->time->delta_time;
+	}
+}
+
+void Biter::UpdateOffset()
+{
+	offset = float3(0, 0.5 + math::Sin(App->time->time / 300) / 5 + math::Sin(App->time->time / 150) / 20, 0);
+	float3 movement = current_velocity;
+	movement = movement.Cross(float3(0, 1, 0));
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:394]: instead of float3(0,1,0)) use float3::unitY

File: Scripts/src/Script/Biter.cpp [hunk 6]
```diff
@@ -1,396 +1,396 @@
...
+	}
+	owner->transform.SetGlobalMatrixTranslation(current_position + offset);
+	collider->UpdateDimensions();
+}
+
+void Biter::MoveIntoDirection(float3 direction, float speed)
+{
+	float3 normalized_direction = direction.Normalized();
+	ComputeHorizontalVelocity(float3(normalized_direction.x, 0.f, normalized_direction.z), speed);
+	//ComputeVerticalVelocity for charge -> posible TO-DO
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:303]: remove if not used

File: Scripts/src/Script/Biter.cpp [hunk 7]
```diff
@@ -1,396 +1,396 @@
...
+
+	if (next_position.Equals(float3(0, 0, 0)) || App->time->time > time_count_delay)
+	{
+		next_position = target;
+		time_count_delay = App->time->time + rand() % 500 + 300;
+	}
+	AirMoveTo(next_position);
+
+	float distance = current_position.Distance(target);
+	if (distance <= explosion_radius && abs(current_position.y - target.y) < 10.0F && explosion_current_cooldown <= 0)
```
Comment (JorxSS) [inline Scripts/src/Script/Biter.cpp:238]: 10.f is a magic number? make it as a member of the class

**Review by JorxSS — CHANGES_REQUESTED** [review]
Check the comments and do the changes, it's quite good btw!!

**Review by Dartemiss — APPROVED** [review]
Seems correct! GJ

**Comment by Pacasasgar** [comment]
> Check the comments and do the changes, it's quite good btw!!

Done!

**Review by JorxSS — APPROVED** [review]
GJ

**Pull request merged**